### PR TITLE
Override all `Read` methods in `NullTextReader` and `NullStreamReader`

### PR DESCRIPTION
**Pull request by bgrainger**
This is a fast-follow to https://github.com/dotnet/runtime/pull/61898#discussion_r789742168:

> There are a few other TextReader-derived types in dotnet/runtime, e.g. a couple in System.Console, and NullTextReader in corelib. We should likely override the new overloads on those as well. (I'm not sure why NullTextReader doesn't already override most of the virtuals, but it seems it should, for perf.) Same goes for NullStreamReader.

CC @adamsitnik

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs [hunk 1]
```diff
@@ -1395,35 +1395,43 @@ protected override void Dispose(bool disposing)
...
+
+            public override ValueTask<string?> ReadLineAsync(CancellationToken cancellationToken) =>
+                cancellationToken.IsCancellationRequested ? ValueTask.FromCanceled<string?>(cancellationToken) : default;
+
+            public override string ReadToEnd() => "";
+
+            public override Task<string> ReadToEndAsync() => Task.FromResult("");
+
+            public override Task<string> ReadToEndAsync(CancellationToken cancellationToken) =>
+                cancellationToken.IsCancellationRequested ? Task.FromCanceled<string>(cancellationToken) : Task.FromResult("");
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs:1432]: We should cache `Task.FromResult("")` into a static field (maybe on https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCache.cs) and return that cached object, as FromResult doesn't cache a task for an empty string.

File: src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs [hunk 2]
```diff
@@ -342,15 +342,41 @@ private sealed class NullTextReader : TextReader
         {
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:342]: I wonder if we even need NullTextReader?  Seems like `public static readonly TextReader Null = new NullTextReader();` could be changed to instead construct a `NullStreamReader`.  It's possible there's some reason I'm missing not to do that, but it'd be nice to avoid the duplication if possible. @adamsitnik, thoughts?

Comment (bgrainger) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:342]: I wondered this, too, as I refactored the classes to be almost identical; the only thing I can think of changing/breaking is code that (essentially) performs `TextReader.Null is TextReader and not StreamReader` (or similar `switch`/`case` or `if`/`else`).

If setting `TextReader.Null` to a `NullStreamReader` (and deleting `NullTextReader`) is desirable, let me know and I'm happy to make the change.

File: src/libraries/System.IO/tests/Stream/Stream.NullTests.cs [hunk 3]
```diff
@@ -131,16 +131,57 @@ public static void TestNullTextReader(TextReader input)
...
+            Assert.Equal(-1, input.Read());
+            Assert.Equal(-1, input.Peek());
+
+            var chars = new char[2];
+            Assert.Equal(0, input.Read(chars, 0, chars.Length));
+            Assert.Equal(0, input.Read(chars.AsSpan()));
+            Assert.Equal(0, input.ReadBlock(chars, 0, chars.Length));
+            Assert.Equal(0, input.ReadBlock(chars.AsSpan()));
+            Assert.Equal("", input.ReadToEnd());
+            input.Dispose();
```
Comment (adamsitnik) [inline src/libraries/System.IO/tests/Stream/Stream.NullTests.cs:149]: the input has been already disposed. In case you want to test whether `Dispose()` can be called multiple times please add a separate test for that (it would be more obvious).
```suggestion
```

File: src/libraries/System.IO/tests/Stream/Stream.NullTests.cs [hunk 4]
```diff
@@ -131,16 +131,57 @@ public static void TestNullTextReader(TextReader input)
...
+            Assert.Equal("", await input.ReadToEndAsync());
+            Assert.Equal("", await input.ReadToEndAsync(default));
+            input.Dispose();
+        }
+
+        [Theory]
+        [MemberData(nameof(NullReaders))]
+        public static async Task TestCanceledNullTextReaderAsync(TextReader input)
+        {
+            using var tokenSource = new CancellationTokenSource();
```
Comment (adamsitnik) [inline src/libraries/System.IO/tests/Stream/Stream.NullTests.cs:172]: nit: there is no need to use `var`

```suggestion
            using CancellationTokenSource tokenSource = new ();
```

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCache.cs [hunk 5]
```diff
@@ -13,6 +13,8 @@ internal static class TaskCache
         internal static readonly Task<bool> s_trueTask = CreateCacheableTask(result: true);
         /// <summary>A cached Task{Boolean}.Result == false.</summary>
         internal static readonly Task<bool> s_falseTask = CreateCacheableTask(result: false);
+        /// <summary>A cached Task{String}.Result == "".</summary>
+        internal static readonly Task<string> s_emptyStringTask = CreateCacheableTask(result: "");
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCache.cs:17]: We optimize for common scenarios. I don't think that reading from `xReader.Null` is common and it should not be on a hot path in a well designed application.

Caching tasks like that is not for free (it prolongs the time it takes to initialize the static fields), please remove the cached task.

File: src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs (hunk 2 above)
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:342]: The scenario provided by @bgrainger is valid, but I hope that nobody does that.

I've taken a look at how these two properties are [typically used](https://cs.github.com/?q=StreamReader.Null%20language%3AC%23) and I think that it's relatively safe to reduce the code duplication and remove `NullTextReader`.

**Review by adamsitnik — COMMENTED** [review]
Thank you for your contribution @bgrainger ! PTAL at my comments.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCache.cs (hunk 5 above)
Comment (bgrainger) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCache.cs:17]: I'm assuming you'd prefer to go back to `Task.FromResult("")` in the method body, not even using a `static readonly` field in `NullStreamReader` itself; I will revert this change (which was suggested at https://github.com/dotnet/runtime/pull/64301/files/426c0414a3a7ad2759a44acb9fd5abbb9c6505e6#r792278679).

File: src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs [hunk 6]
```diff
@@ -19,7 +19,7 @@ namespace System.IO
     // There are methods on the Stream class for reading bytes.
     public abstract partial class TextReader : MarshalByRefObject, IDisposable
     {
-        public static readonly TextReader Null = new NullTextReader();
+        public static readonly TextReader Null = StreamReader.Null;
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:22]: Overall the code looks good, but for some reason the tests are failing on **mono** (the CI has not finished so I am not 100% sure that it's mono bug):

```log
  Discovering: System.IO.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.IO.Tests (found 694 of 700 test cases)
  Starting:    System.IO.Tests (parallel test collections = on, max threads = 6)
    System.IO.Tests.NullTests.TestCanceledNullTextReaderAsync(input: null) [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.OperationCanceledException)
      Actual:   typeof(System.NullReferenceException): Object reference not set to an instance of an object
      ---- System.NullReferenceException : Object reference not set to an instance of an object
      Stack Trace:
        /_/src/libraries/System.IO/tests/Stream/Stream.NullTests.cs(184,0): at System.IO.Tests.NullTests.<>c__DisplayClass11_0.<<TestCanceledNullTextReaderAsync>b__0>d.MoveNext()
        --- End of stack trace from previous location ---
        ----- Inner Stack Trace -----
        /_/src/libraries/System.IO/tests/Stream/Stream.NullTests.cs(184,0): at System.IO.Tests.NullTests.<>c__DisplayClass11_0.<<TestCanceledNullTextReaderAsync>b__0>d.MoveNext()
        --- End of stack trace from previous location ---
      Assert.Throws() Failure
```

Is there any chance you could build it and try to repro locally? https://github.com/dotnet/runtime/blob/main/docs/workflow/building/mono/README.md

I have no idea why it fails only on Mono, but I guess it's caused by the fact that two out of three arugments provided by `NullReaders` refer to the same object instance.


https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.IO/tests/Stream/Stream.NullTests.cs#L197-L203

Comment (bgrainger) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:22]: I believe it's actually 3 out of 3 now. `StringReader.Null` always inherited from the base class, so it was the same object as `TextReader.Null`. But now they all are the same object. 

I could throw on a `.Distinct()` to ensure that all the properties are read, but that only unique objects are tested, but I'm not sure if that would stop `null` being passed as the input to the test method.

Comment (bgrainger) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:22]: Just had a thought that perhaps this is a static field initialisation problem (only on mono)? If the base class static fields have to be initialised before the derived class, then `TextReader.Null` would be set to `null`, which would cause the test failure. The fix would be to make `NullStreamReader` internal and simply construct a second instance in `TextReader.Null`. (`StreamReader.Null` technically wouldn't need to be a second instance, but could be `= (StreamReader) TextReader.Null` but that feels unnecessarily brittle to me.)

Comment (bgrainger) [reply src/libraries/System.Private.CoreLib/src/System/IO/TextReader.cs:22]: > I could throw on a `.Distinct()`

That didn't fix the broken builds; I'll move ahead with the static field theory. (The `Distinct()` is still useful to avoid unnecessary testing.)

**Comment by bgrainger** [comment]
Changing the static field initialisation has resolved the "Libraries Test Run release mono" build failures.

File: src/libraries/System.IO/tests/Stream/Stream.NullTests.cs [hunk 7]
```diff
@@ -194,15 +242,10 @@ public void DisposeAsync_Nop()
...
-            {
-                yield return new object[] { TextReader.Null };
-                yield return new object[] { StreamReader.Null };
-                yield return new object[] { StringReader.Null };
-            }
-        }
+        public static IEnumerable<object[]> NullReaders =>
+            new[] { TextReader.Null, StreamReader.Null, StringReader.Null }
+                .Distinct()
+                .Select(x => new object[] { x });
```
Comment (adamsitnik) [inline src/libraries/System.IO/tests/Stream/Stream.NullTests.cs:248]: can we revert these changes now?

```suggestion
        public static IEnumerable<object[]> NullReaders
        {
            get
            {
                yield return new object[] { TextReader.Null };
                yield return new object[] { StreamReader.Null };
                yield return new object[] { StringReader.Null };
            }
        }
```

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @bgrainger !
